### PR TITLE
Adding Sentence Order Prediction [WIP]

### PR DESCRIPTION
**Pull request by pruksmhc**
Adding Sentence Order Prediction Task for ALBERT
What this version of MLM supports: ALBERT embedder. 
Below are the runs for MLM + SOP + intermediate task (which is how we intend to use SOP). 

| Task      | Task performance (best ALBERT finetuned on task) |  MLM perplexity | SOP performance|
| ------------- |:-------------:| -------------:|  -------------:| 
| CCG   | 0.944 |5.123   | 0.978 |
| QQP     | 0.8594    | 4.303 |0.8428 |

### CODE REVIEW

**Comment by sleepinyourhat** [comment]
That was fast! Taking a look shortly...

File: jiant/models.py [hunk 1]
```diff
@@ -186,7 +187,6 @@ def build_sent_encoder(args, vocab, d_emb, tasks, embedder, cove_layer):
             sep_embs_for_skip=args.sep_embs_for_skip,
             cove_layer=cove_layer,
         )
-        d_sent = 2 * args.d_hid
```
Comment (sleepinyourhat) [inline jiant/models.py:189]: What's this for?

File: jiant/models.py [hunk 2]
```diff
@@ -541,6 +541,9 @@ def build_task_specific_modules(task, model, d_sent, d_emb, vocab, embedder, arg
             params=task_params,
         )
         setattr(model, "%s_mdl" % task.name, module)
+    elif isinstance(task, SentenceOrderTask):
```
Comment (sleepinyourhat) [inline jiant/models.py:544]: Is there a reason this isn't a PairClassificationTask? In any case, it probably makes more sense readability-wise to cluster this with the single-purpose task classes below, rather than with the broad classes here. (Just proposing to rearrange the else clauses, not change behavior.)

File: jiant/models.py [hunk 3]
```diff
@@ -687,7 +690,33 @@ def build_single_sentence_module(task, d_inp: int, project_before_pooling: bool,
...
+    # exposed in huggingface for the pretrained weights, we set projection to true.
+    project_before_pooling = True
+    module = build_pair_sentence_module(task, d_inp, model, params, project_before_pooling)
+    # The huggingface implementation exposes the pretrained projection layer for the SOP task, which
+    # we use. See: https://github.com/huggingface/transformers/issues/2671 for more details.
+    module.pooler.project = model.sent_encoder._text_field_embedder.model.pooler
+    return module
+
+
+def build_pair_sentence_module(task, d_inp, model, params, project_before_pooling=None):
```
Comment (sleepinyourhat) [inline jiant/models.py:719]: `project_before_pooling=None` appears to be new. Why?

Comment (sleepinyourhat) [reply jiant/models.py:719]: Similar questions for the changes elsewhere in this fn.

File: jiant/models.py [hunk 4]
```diff
@@ -687,7 +690,33 @@ def build_single_sentence_module(task, d_inp: int, project_before_pooling: bool,
     return module
 
 
-def build_pair_sentence_module(task, d_inp, model, params):
+def build_sop(task, d_inp, model, params):
+    """
+    Build and load the pretrained head for the sentence order prediction task.
```
Comment (sleepinyourhat) [inline jiant/models.py:695]: Mention ALBERT support here?

**Comment by sleepinyourhat** [comment]
General sanity check before we run anything big/expensive:
- Make sure that we're getting better-than-chance SOP accuracy out of the box with ALBERT, before any training.

File: jiant/models.py (hunk 1 above)
Comment (pruksmhc) [reply jiant/models.py:189]: Ah this is an unnecessary change and nothing to do with this PR. I will revert.

File: jiant/models.py [hunk 5]
```diff
@@ -687,7 +690,33 @@ def build_single_sentence_module(task, d_inp: int, project_before_pooling: bool,
...
+    model: MultiTaskModel,
+    params: Params
+
+    Returns
+    -------
+    module: Pooler object, which is loaded with pretrained weights from ALBERT SOP
+    pretraining. 
+
+    """
+    # By default, we do not project before pooling. Since the projection weights is what is
```
Comment (sleepinyourhat) [inline jiant/models.py:709]: nit: 'weights is'

File: jiant/models.py [hunk 6]
```diff
@@ -1097,6 +1127,7 @@ def _tagger_forward(self, batch: dict, task: TaggingTask, predict: bool) -> dict
         """
         out = {}
         # batch[inputs] only has one item
+
```
Comment (pruksmhc) [inline jiant/models.py:1130]: Will revert these extra spaces.

File: jiant/models.py [hunk 7]
```diff
@@ -687,7 +690,33 @@ def build_single_sentence_module(task, d_inp: int, project_before_pooling: bool,
...
+    params: Params
+
+    Returns
+    -------
+    module: Pooler object, which is loaded with pretrained weights from ALBERT SOP
+    pretraining. 
+
+    """
+    # By default, we do not project before pooling. Since the projection weights is what is
+    # exposed in huggingface for the pretrained weights, we set projection to true.
```
Comment (sleepinyourhat) [inline jiant/models.py:710]: Clarify: Is this a projection _before pooling_ , or something else? I'm not aware of any kind of pooling operation in ALBERT.

File: jiant/tasks/tasks.py [hunk 8]
```diff
@@ -3731,3 +3732,121 @@ def get_metrics(self, reset=False):
         """Get metrics specific to the task"""
         acc = self.scorer1.get_metric(reset)
         return {"accuracy": acc}
+
+
+@register_task("wikipedia_corpus_sop", rel_path="wikipedia_corpus_small")
+class SentenceOrderTask(PairClassificationTask):
+    """ Task class for Sentence Order Prediction (SOP) with wikipedia data.
+        See the ALBERT paper for details on SOP: https://arxiv.org/abs/1909.11942.
+        We are currently using an unpreprocessed version of the Wikipedia corpus
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:3741]: Say a tiny bit more about the data source here, even if the full docs live elsewhere—there is no one 'Wikipedia corpus'.

File: jiant/models.py (hunk 2 above)
Comment (sleepinyourhat) [reply jiant/models.py:545]: Ah, I see that this is a PairClassificationTask, but it's the only one that has to preempt the generic build_sentence_pair_module, so it has to be here. This seems like it's making an already messy pile of if-elses a bit harder to follow. I don't see any obvious alternative that doesn't involve a larger refactor, but it'd be good to get @pyeres's input before a merge.

File: jiant/tasks/tasks.py [hunk 9]
```diff
@@ -3731,3 +3732,121 @@ def get_metrics(self, reset=False):
...
+        """Loading data file and tokenizing the text
+        Args:
+            path: (str) data file path
+        """
+        count = 0
+        with open(path, encoding="utf-8") as txt_fh:
+
+            for row in txt_fh:
+                count += 1
+                if count == 10:
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:3770]: What's special about 10? Add a comment.

File: jiant/tasks/tasks.py [hunk 10]
```diff
@@ -3731,3 +3732,121 @@ def get_metrics(self, reset=False):
...
+                    break
+                toks = row.strip()
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:3772]: This looks fishy...

File: jiant/tasks/tasks.py [hunk 11]
```diff
@@ -3731,3 +3732,121 @@ def get_metrics(self, reset=False):
...
+                sentences = row.split(".")
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:3773]: This looks very fishy. What about other end-of-sentence punctuation like '!'? What does the ALBERT code do?

Comment (sleepinyourhat) [reply jiant/tasks/tasks.py:3773]: (We should check on this before running any large experiments, since it could hurt performance in a way that's incremental/hard to detect.)

File: jiant/tasks/tasks.py (hunk 10 above)
Comment (sleepinyourhat) [reply jiant/tasks/tasks.py:3772]: It looks like it's never used...

File: jiant/tasks/tasks.py [hunk 12]
```diff
@@ -3731,3 +3732,121 @@ def get_metrics(self, reset=False):
...
+                if len(sentences) <= 1:
+                    continue
+                else:
+                    for i in range(len(sentences) - 1):
+                        if random.uniform(0, 1) > 0.5:
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:3778]: Genuine question: Is there a more idiomatic way to do this? `random` doesn't have a Bernoulli method built-in. Is randint any faster or more standard?

File: jiant/models.py (hunk 7 above)
Comment (pruksmhc) [reply jiant/models.py:711]: What they call a "pooler" is a projection: 
https://github.com/huggingface/transformers/blob/ddb6f9476b58ed9bf4433622ca9aa49932929bc0/src/transformers/modeling_albert.py#L446

File: jiant/tasks/tasks.py [hunk 13]
```diff
@@ -3731,3 +3732,121 @@ def get_metrics(self, reset=False):
...
+            for row in txt_fh:
+                count += 1
+                if count == 10:
+                    break
+                toks = row.strip()
+                sentences = row.split(".")
+                if len(sentences) <= 1:
+                    continue
+                else:
+                    for i in range(len(sentences) - 1):
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:3777]: How does this block of code compare with code in HuggingFace/ALBERT?

File: jiant/models.py (hunk 7 above)
Comment (pruksmhc) [reply jiant/models.py:711]: and this is used here to create the pooled_output (which is the head used in the huggingface code for SOP prediction) https://github.com/huggingface/transformers/blob/ddb6f9476b58ed9bf4433622ca9aa49932929bc0/src/transformers/modeling_albert.py#L570

File: jiant/tasks/tasks.py [hunk 14]
```diff
@@ -3731,3 +3732,121 @@ def get_metrics(self, reset=False):
...
+                        )
+                        yield (sent_a_processed, sent_b_processed, is_right_order)
+
+    def load_data(self):
+        pass
+
+    def process_split(
+        self, split, indexers, model_preprocessing_interface
+    ) -> Iterable[Type[Instance]]:
+        """Process a sentence order prediction split by indexing and creating fields.
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:3800]: Clarify why this needs to override `process_split`, when most pair tasks don't. I assume it has something to do with making sure the label names match what's loaded from pretrained ALBERT? (This kind of information should be helpful for future cleanup of the structure of tasks...)

File: jiant/tasks/tasks.py [hunk 15]
```diff
@@ -3731,3 +3732,121 @@ def get_metrics(self, reset=False):
...
+        Args:
+            split: (list) a single list of sentences
+            indexers: (Indexer object) indexer to index input words
+        """
+
+        def _make_instance(sent_):
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:3806]: I don't have much context for reading this part of the task logic. @W4ngatang? @HaokunLiu?

File: jiant/models.py (hunk 7 above)
Comment (pruksmhc) [reply jiant/models.py:711]: basically, what the huggingface implementation of ALBERT does is, for the SOP head, takes the output of the ALBERT encoding and passes it through a linear layer and an activation function. The closest thing we have to that is this project_before_pooling attribute in the PairClassifier.

File: jiant/models.py (hunk 3 above)
Comment (pruksmhc) [reply jiant/models.py:720]: This is because I wanted to be able to set project_before_pooling=True for this experiment (see comment in previous thread about why), but here it was set to False for transformer based models It seems as though you were the one who added this -why was this added? https://github.com/nyu-mll/jiant/blob/master/jiant/models.py#L839

File: jiant/tasks/tasks.py [hunk 16]
```diff
@@ -3731,3 +3732,121 @@ def get_metrics(self, reset=False):
...
+                    model_preprocessing_interface.boundary_token_fn(sent_b), indexers
+                )
+                label = LabelField(is_right_order, label_namespace="labels", skip_indexing=True)
+                d = {"input1": inp1, "input2": inp2, "targs": label}
+            return Instance(d)
+
+        for sent in split:
+            yield _make_instance(sent)
+
+    def count_examples(self):
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:3827]: Same here—not sure what the history/intent is behind when this is/isn't overridden.

File: jiant/tasks/tasks.py [hunk 17]
```diff
@@ -3731,3 +3732,121 @@ def get_metrics(self, reset=False):
...
+        """Computes number of samples
+        Assuming every line is one example.
+        """
+        example_counts = {}
+        for split, split_path in self.files_by_split.items():
+            example_counts[split] = sum(1 for _ in self.get_data_iter(split_path))
+        self.example_counts = example_counts
+
+    def get_split_text(self, split: str):
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py:3836]: Same as above for the next two methods.

File: jiant/models.py (hunk 7 above)
Comment (sleepinyourhat) [reply jiant/models.py:711]: Ah, got it. This could use a clarification in the code comments here, though it doesn't help that I've lost track of where our pooling logic lives.

File: jiant/models.py (hunk 3 above)
Comment (sleepinyourhat) [reply jiant/models.py:720]: There's no reason to project (with a linear layer) _before_ pooling with BERT-style models, where pooling is just a first- or last-token selector. This kind of projection only really makes sense with max-pooling.

File: jiant/tasks/tasks.py (hunk 11 above)
Comment (pruksmhc) [reply jiant/tasks/tasks.py:3782]: @phu-pmh

**Comment by pruksmhc** [comment]
I'm reworking this task entirely (basically not making it inherit from Pairclassificationtask) to make the code most cleaner/clearer, so the game plan is to incorporate comments from here. I'm going to start another PR, since most of these comments will be resolved with this change.

File: jiant/tasks/tasks.py (hunk 12 above)
Comment (pruksmhc) [reply jiant/tasks/tasks.py:3787]: Hmm I'm not sure. Taking a look.

File: jiant/tasks/tasks.py (hunk 13 above)
Comment (pruksmhc) [reply jiant/tasks/tasks.py:3786]: SOP isn't implemented in Huggingface.

**Comment by RobinAlgayres** [comment]
Hi, is there a way to use the Sentence Order Prediction currently? From jiant or from huggingface? 
Thank you

**Comment by jeswan** [comment]
@RobinAlgayres This is currently not supported in jiant v2. What tasks/models are you planning to use with Sentence Order Prediction. We can help point you in the right direction.

**Comment by RobinAlgayres** [comment]
Well I would like to use a BERT model pretrained on SOP and use it on a dataset made of pairs of sentences. I do not wish to do finetuning so the model should output a boolean somewhere when I give it a pair of sentences and directly tell me if those sentences are following each other. 
For example, the BertForNextSentencePredictio from huggingface does that but it's trained on NSP.
Do you have any ideas?
